test/server/connectors/serviceInsights/detectCycles.spec.js

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 Expedia Group
+ *
+ *       Licensed under the Apache License, Version 2.0 (the License);
+ *       you may not use this file except in compliance with the License.
+ *       You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *       Unless required by applicable law or agreed to in writing, software
+ *       distributed under the License is distributed on an AS IS BASIS,
+ *       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *       See the License for the specific language governing permissions and
+ *       limitations under the License.
+ *
+ */
+
+import {expect} from 'chai';
+
+const {detectCycles} = require('../../../../server/connectors/serviceInsights/detectCycles');
+
+function getBaseGraph(primaryIdKey = 'id') {
+    return {
```

Comment:
thanks for this example, it helps to show the data model. I was wondering what impact there would be to retaining trace IDs in the results here.. this would allow you to tell causally which paths were caused by an upstream vs originating mid-tier. I expect since the api results are only 10 or 1000 traces, this could be practical to do client side. We were thinking of allowing navigation based on this in zipkin, though similar to you starting with just links in efforts towards incremental progress

Reply:
We are looking into this as a feature to be developed. Hovering over a link could show the traceIds or a user could search for a specific traceId(s) from which we can build the graph.